src/msg/simple/Pipe.cc

```diff
@@ -2601,10 +2601,21 @@ int Pipe::tcp_read_wait()
 
 ssize_t Pipe::do_recv(char *buf, size_t len, int flags)
 {
+  int retries = 0;
 again:
   ssize_t got = ::recv( sd, buf, len, flags );
   if (got < 0) {
-    if (errno == EAGAIN || errno == EINTR) {
+    if (errno == EINTR) {
+      goto again;
+    } else if (errno == EAGAIN &&
+	       msgr->cct->_conf->ms_socket_retry_on_eagain > retries) {
```

Comment:
The are 2 reasons why recvfrom/recv can return EAGAIN:

1. receive timeout (previously set with setsockopt) has expired
2. there was no data in the socket, and non-blocking operation has been requested

If the timeout has expired one should return the error (retrying denies the purpose of setting a timeout). On the other hand, waiting for data to arrive in a busy loop makes little sense (most likely it won't arrive after 2 repetitions)